### PR TITLE
[WIP][HIVEMALL-24] Scalable field-aware factorization machines

### PR DESCRIPTION
**Pull request by takuti**
## What changes were proposed in this pull request?

Make `ffm_predict` function more scalable by creating its UDAF implementation

## What type of PR is it?

Improvement

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-24

## How was this patch tested?

Not yet

## How to use this feature?

https://gist.github.com/takuti/c49dfe2d06cb12bf69bad30213d0afc3

### CODE REVIEW

**Comment by takuti** [comment]
### feature_pairs()

```sql
drop temporary function if exists feature_pairs;
create temporary function feature_pairs as 'hivemall.ftvec.pairing.FeaturePairsUDTF';

select
  t1.rowid,
  t2.i,
  t2.j,
  t2.Xi,
  t2.Xj
from (
  select 1 as rowid, array("1:3:1", "1:11:2", "2:14:3", "3:19:4") as features
) t1
LATERAL VIEW feature_pairs(features, "-ffm -feature_hashing 10 -num_fields 5") t2 as i, j, Xi, Xj;
```

Output:

```
t1.rowid        t2.i    t2.j    t2.xi   t2.xj
1       0       NULL    NULL    NULL
1       0       NULL    1.0     NULL
1       16      56      1.0     2.0
1       17      71      1.0     3.0
1       18      96      1.0     4.0
1       1       NULL    2.0     NULL
1       57      71      2.0     3.0
1       58      96      2.0     4.0
1       2       NULL    3.0     NULL
1       73      97      3.0     4.0
1       3       NULL    4.0     NULL
```

Todo:

- Support a case that `field` is not specified e.g., `11:2`

**Comment by myui** [comment]
```sql
WITH testing_fm_exploded as (
  select
    t1.rowid, 
    t2.i,
    t2.j,
    t2.Xi,
    t2.Xj
  from
    testing t1
    LATERAL VIEW ffm_pairs(features) t2 as i, j, Xi, Xj
)
select
  ffm_predict( -- UDAF
    t1.Xi,
    t1.Xj,
    p1.Wi,
    p1.Vi, -- Vij
    p2.Vi  -- Vji
  ) as predicted
from 
  testing_fm_exploded t1
  LEFT OUTER JOIN fm_model p1 ON (p1.i = t1.i)
  LEFT OUTER JOIN fm_model p2 ON (p2.i = t1.j and p2.modelid = p1.modelid)  
```

File: core/src/main/java/hivemall/fm/FFMPredictGenericUDAF.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+            }
+        }
+
+        @Override
+        public Object terminatePartial(@SuppressWarnings("deprecation") AggregationBuffer agg)
+                throws HiveException {
+            FFMPredictAggregationBuffer myAggr = (FFMPredictAggregationBuffer) agg;
+
+            final Object[] partialResult = new Object[1];
+            return partialResult;
```
Comment (myui) [inline core/src/main/java/hivemall/fm/FFMPredictGenericUDAF.java]: partialResult is not filled..

File: core/src/main/java/hivemall/fm/FFMPredictGenericUDAF.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+            // initialize input
+            if (mode == Mode.PARTIAL1 || mode == Mode.COMPLETE) {// from original data
+                this.wiOI = HiveUtils.asDoubleCompatibleOI(parameters[0]);
+                this.vijOI = HiveUtils.asListOI(parameters[1]);
+                this.vjiOI = HiveUtils.asListOI(parameters[2]);
+                this.xiOI = HiveUtils.asDoubleCompatibleOI(parameters[3]);
+                this.xjOI = HiveUtils.asDoubleCompatibleOI(parameters[4]);
+            } else {// from partial aggregation
+                StructObjectInspector soi = (StructObjectInspector) parameters[0];
```
Comment (myui) [inline core/src/main/java/hivemall/fm/FFMPredictGenericUDAF.java]: doubleOI is enough

File: core/src/main/java/hivemall/fm/FFMPredictGenericUDAF.java [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+                this.internalMergeOI = soi;
+                this.sumField = soi.getStructFieldRef("sum");
+            }
+
+            // initialize output
+            final ObjectInspector outputOI;
+            if (mode == Mode.PARTIAL1 || mode == Mode.PARTIAL2) {// terminatePartial
+                outputOI = internalMergeOI();
```
Comment (myui) [inline core/src/main/java/hivemall/fm/FFMPredictGenericUDAF.java]: doubleOI is enough.

**Comment by myui** [comment]
I'm working on this issue in https://github.com/apache/incubator-hivemall/pull/105

**Pull request closed without merging**